cmd/kube-spawn/kube-spawn.go

```diff
@@ -20,40 +20,81 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"path/filepath"
 
+	"github.com/kinvolk/kube-spawn/pkg/utils"
 	"github.com/spf13/cobra"
 )
 
 const k8sStableVersion string = "1.7.0"
+const defaultRuntime string = "docker"
+const kubeSpawnDirDefault string = "/var/lib/kube-spawn"
 
 var (
 	cmdKubeSpawn = &cobra.Command{
 		Use:   "kube-spawn",
 		Short: "kube-spawn is a tool for creating a multi-node dev Kubernetes cluster",
 		Long:  "kube-spawn is a tool for creating a multi-node dev Kubernetes cluster, by using the local source code and systemd-nspawn containers",
-		Run: func(cmd *cobra.Command, args []string) {
-			if printVersion {
-				fmt.Printf("kube-spawn %s\n", version)
-				os.Exit(0)
-			}
-			if err := cmd.Usage(); err != nil {
-				log.Fatal(err)
-			}
-		},
+		Run:   runUsage,
 	}
 
 	version      string
 	k8srelease   string
+	k8sruntime   string
+	rktBinDir    string
+	rktletBinDir string
 	printVersion bool
+	kubeSpawnDir string
+
+	kubeadmCgroupDriver     string
+	kubeadmRuntimeEndpoint  string
+	kubeadmRequestTimeout   string = "15m"
+	kubeadmContainerRuntime string
+
+	stderr *log.Logger
+	stdout *log.Logger
 )
 
+func runUsage(cmd *cobra.Command, args []string) {
+	if printVersion {
+		stdout.Printf("kube-spawn %s\n", version)
+		os.Exit(0)
+	}
+	if err := cmd.Usage(); err != nil {
+		stderr.Fatal(err)
+	}
+}
+
 func init() {
-	cmdKubeSpawn.Flags().BoolVarP(&printVersion, "version", "V", false, "output version information")
+	cmdKubeSpawn.Flags().BoolVarP(&printVersion, "version", "V", false, "Output version information")
+	cmdKubeSpawn.PersistentFlags().StringVarP(&k8sruntime, "kubernetes-runtime", "r", defaultRuntime, "Runtime to use for the spawned cluster (docker or rkt)")
+	cmdKubeSpawn.PersistentFlags().StringVar(&rktBinDir, "rkt-bin-dir", "", "path to rkt binaries")
+	cmdKubeSpawn.PersistentFlags().StringVar(&rktletBinDir, "rktlet-bin-dir", "", "path to rktlet binaries")
 	cmdKubeSpawn.PersistentFlags().StringVarP(&k8srelease, "kubernetes-version", "k", k8sStableVersion, "Kubernetes version to spawn, \"\" or \"dev\" for self-building upstream K8s.")
+	cmdKubeSpawn.PersistentFlags().StringVarP(&kubeSpawnDir, "kube-spawn-dir", "d", kubeSpawnDirDefault, "path to kube-spawn asset directory")
+
+	cmdKubeSpawn.PersistentPreRun = func(cmd *cobra.Command, args []string) {
+		stderr = log.New(os.Stderr, fmt.Sprintf("[%v] ", cmd.Name()), log.Lshortfile)
+		stdout = log.New(os.Stdout, "", 0)
```

Comment:
What's the reasoning behind this?

Reply:
Having info and error output rather than printing everything as a log with date/time

Reply:
I meant, why do you want some messages go to stdout instead of stderr? Why do messages going to stdout not have the `[cmd]` prefix? If it's just for the format, a `log.SetPrefix(fmt.Sprintf("[%v] ", cmd.Name()))` and `log.SetFlags(log.Lshortfile)` would do as well. What is with all the messages from `pkg/...`?

IMHO, if you want to have "better" logging, try to write down a design first, as far as possible something which suffices both current (again, what about all the code in `pkg`?) and future log usage. And use one of the established libraries in Go (or, if really necessary, build a custom logger pkg compatible with the standard library logger).

Right now it's a wild mix, even in the main package (`git grep 'log\.' cmd/kube-spawn`).

Reply:
I agree 100%
I started working on the output stuff in a separate commit that should have been excluded from this PR in the first place...

Reply:
I will open a PR for the logging and output thing.